File: src/main/java/io/github/binaryfoo/decoders/EmvBitStringDecoder.java
package io.github.binaryfoo.decoders;

import io.github.binaryfoo.DecodedData;
import io.github.binaryfoo.Decoder;
import io.github.binaryfoo.bit.BitPackage;
import io.github.binaryfoo.bit.EmvBit;
import io.github.binaryfoo.decoders.bit.BitStringField;
import io.github.binaryfoo.decoders.bit.EmvBitStringParser;
import io.github.binaryfoo.io.ClasspathIO;
import io.github.binaryfoo.tlv.ISOUtil;
import org.apache.commons.io.IOUtils;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class EmvBitStringDecoder implements Decoder {

    private final List<BitStringField> bitMappings;
    private final int lengthInCharacters;
    private final boolean showFieldHexInDecode;

    public EmvBitStringDecoder(String fileName, boolean showFieldHexInDecode) {
        this(ClasspathIO.open(fileName), showFieldHexInDecode);
    }

    public EmvBitStringDecoder(InputStream input, boolean showFieldHexInDecode) {
        this.showFieldHexInDecode = showFieldHexInDecode;
        try {
            bitMappings = EmvBitStringParser.parse(IOUtils.readLines(input));
        } catch (IOException e) {
            throw new RuntimeException(e);
        } finally {
            IOUtils.closeQuietly(input);
        }
        this.lengthInCharacters = findMaxLengthInBytes() * 2;
    }

    private int findMaxLengthInBytes() {
        int maxLength = 0;
        for (BitStringField mapping : bitMappings) {
            maxLength = Math.max(mapping.getStartBytesOffset() + mapping.getLengthInBytes(), maxLength);
        }
        return maxLength;
    }

    @Override
    public List<DecodedData> decode(@NotNull String input, int startIndexInBytes, @NotNull DecodeSession decodeSession) {
        List<DecodedData> decoded = new ArrayList<>();
        Set<EmvBit> bits = BitPackage.fromHex(input);
        for (BitStringField field : bitMappings) {
            String v = field.getValueIn(bits);
            if (v != null) {
                int fieldStartIndex = startIndexInBytes + field.getStartBytesOffset();
                decoded.add(DecodedData.primitive(field.getPositionIn(showFieldHexInDecode ? bits : null), v, fieldStartIndex, fieldStartIndex + field.getLengthInBytes()));
            }
        }
        return decoded;
    }

    @Override
    public String validate(String bitString) {
        if (bitString == null || bitString.length() != lengthInCharacters) {
            return String.format("Value must be exactly %d characters", lengthInCharacters);
        }
        if (!ISOUtil.isValidHexString(bitString)) {
            return "Value must contain only the characters 0-9 and A-F";
        }
        return null;
    }

    @Override
    public int getMaxLength() {
        return lengthInCharacters;
    }

}

File: src/main/java/io/github/binaryfoo/decoders/EmvBitStringDecoder.kt
package io.github.binaryfoo.decoders

import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.Decoder
import io.github.binaryfoo.bit.fromHex
import io.github.binaryfoo.decoders.bit.BitStringField
import io.github.binaryfoo.decoders.bit.EmvBitStringParser
import io.github.binaryfoo.res.ClasspathIO
import io.github.binaryfoo.tlv.ISOUtil
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.util.*

/**
 * Decoder based on the config language (DSL...).
 * <p>
 * Each line in fileName should be parseable in one of the 3 formats:
 * <ul>
 *     <li>Enumerated - a direct mapping from a set of bits to a name. Bits can be spread over multiple bytes.</li>
 *     <li>Full byte - similar to enumerated but only handles a single byte.</li>
 *     <li>Numeric - the left or right nibble in a byte is interpreted as an integer.</li>
 * </ul>
 */
open class EmvBitStringDecoder(fileName: String, val showFieldHexInDecoding: Boolean) : Decoder {

    private val bitMappings: List<BitStringField>
    private val maxLength: Int

    init {
        val input = ClasspathIO.open(fileName)
        try {
            bitMappings = EmvBitStringParser.parse(IOUtils.readLines(input))
        } catch (e: IOException) {
            throw RuntimeException(e)
        } finally {
            IOUtils.closeQuietly(input)
        }
        this.maxLength = findMaxLengthInBytes() * 2
    }

    override fun getMaxLength(): Int = maxLength

    private fun findMaxLengthInBytes(): Int {
        if (bitMappings.isEmpty()) {
            return 0
        }
        fun max(a: Int, b: Int): Int = if ((a >= b)) a else b
        return bitMappings.map { it.getStartBytesOffset() + it.getLengthInBytes() }.reduce(::max)
    }

    override fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): List<DecodedData> {
        val decoded = ArrayList<DecodedData>()
        val bits = fromHex(input)
        for (field in bitMappings) {
            val v = field.getValueIn(bits)
            if (v != null) {
                val fieldStartIndex = startIndexInBytes + field.getStartBytesOffset()
                decoded.add(DecodedData.primitive(field.getPositionIn(if (showFieldHexInDecoding) bits else null), v, fieldStartIndex, fieldStartIndex + field.getLengthInBytes()))
            }
        }
        return decoded
    }

    override fun validate(input: String?): String? {
        if (input == null || input.length != maxLength) {
            return "Value must be exactly $maxLength characters"
        }
        if (!ISOUtil.isValidHexString(input)) {
            return "Value must contain only the characters 0-9 and A-F"
        }
        return null
    }

}
